### PR TITLE
Tidy up FailOnTimeoutTest

### PR DESCRIPTION
**Pull request by stefanbirkner**
Improve readability and reusability. When I started to read the test it took me some time to understand it. The reason was the configurable statement `TestStatement` and the `evaluateWith...` methods. The refactored test uses fixed Statements when possible and uses smaller methods for wrapping `FailOnTimeout` with a `ThrowingRunnable` and for wrapping an arbitrary statement with a `FailOnTimeout`. It also calls `FailOnTimeout#evaluate` directly when possible.

### CODE REVIEW

File: src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java
Comment (marcphilipp) [inline src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java:134]: I don't think this has a lot of value since Java does not allow stopping threads anyway. I'm more concerned about all tests that use this leaking threads. Can we tell `InfiniteLoop` instances to stop after we're done testing with them?

**Review by marcphilipp — APPROVED** [review]

File: src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java
Comment (stefanbirkner) [reply src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java]: I replaced InfiniteLoop with a statement that runs for 1 seconds so that it does not run forever. I also rename the test to make clear that it only stops statements that can be interrupted.

**Review by marcphilipp — APPROVED** [review]

File: src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java
Comment (kcooney) [inline src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java:132]: I found this to be confusing. The statement may or may not still be executing. It's not clear here that `RunForASecond` has code that will set it to be true if it is still running after this point. 

I suggest adding a `finally` block in the `RunForASecond` that sets `stillExecuting` to `false`.

Alternatively, replace `stillExecuting` with an `AtomicInteger` and rewrite this to be similar to what it was before.

**Review by kcooney — COMMENTED** [review]
Thanks for the cleanups!

File: src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java
Comment (stefanbirkner) [reply src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java]: Done.

File: src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java
Comment (stefanbirkner) [reply src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java:67]: I thought it is irritating if this test repeats the assertion of the `throwsTestTimedOutException` and that it is also unnecessary because checking the message is independent of the type.
After looking at it again I decided to merge the two tests. I think that the exception type and its message are strongly coupled and therefore the two assertions build once conceptual assertion. Therefore it should be one test instead of two.

File: src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java
Comment (stefanbirkner) [reply src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java:246]: I'm using this class in two tests that call `evaluate` twice and change the delegate between. These tests check that `FailOnTimeout` behaves correctly if it is reused. For example `throwExceptionIfTheSecondCallToEvaluateNeedsTooMuchTime`

    statement.delegate = new FastStatement();
    failOnTimeout.evaluate();
    
    statement.delegate = new RunForASecond();
    assertThrows(
                TestTimedOutException.class,
                run(failOnTimeout));

I'm not completely happy with this but I played around with other ideas and this is the most expressive version for now.

File: src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java
Comment (stefanbirkner) [reply src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java]: I changed the test to use a CountDownLatch: 01dfbc81d49b1c561682119e3eea2c6218b59b44.

File: src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java
Comment (kcooney) [reply src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java]: SGTM. Thanks!

File: src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java
Comment (kcooney) [reply src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java]: Good idea. Perhaps there should still be a short sleep in the `while` loop in `RunForASecond`?

**Review by kcooney — APPROVED** [review]
A few very minor comments, but over all the tests in this class are much better with these changes.

File: src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java
Comment (kcooney) [reply src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java:244]: Actually, making the `volatile` would be a simpler way to address my concerns.

If you are tired of updating this test, I could make these changes in my own pull after your pull is merged. Either way, the test with your changes is much better than it was before.

Comment (stefanbirkner) [reply src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java]: It's fine. I appreciate that you help me to further improve the test. The field is now volatile.

File: src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java
Comment (stefanbirkner) [reply src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java]: It works in the CI and locally. Therefore I would like to go without the sleep until we know about a case where it is necessary.

**Comment by marcphilipp** [comment]
@stefanbirkner Feel free to merge this PR when you're ready.

**Comment by stefanbirkner** [comment]
It is merged. Thank you both for your help.

**Pull request closed without merging**
